### PR TITLE
ace: memory window: Dynamic slots allocation and telemetry base

### PR DESCRIPTION
**Pull request by piotrkmx**
Memory window 2 (aka debug memory window) is divided into 8 pages: 7 slots for different debugging purposes and 1 page containing descriptors of those slots. This PR adds dynamic allocation of memory space for slots.

Implementation of mtrace has been slightly adjusted to work with this mechanism as it was hardcoded before. It just writes data to it's buffer which is later mapped into proper slot. Now, mtrace does not have to handle memory window at all or even know that something like memory window exists.

Telemetry base implementation has been added, which is another piece of code utilizing this mechanism.

CONFIG_DYNAMIC_SLOTS=y
CONFIG_DEBUG_MEMORY_WINDOW=y
CONFIG_TELEMETRY_INTEL=y
CONFIG_LOG_BACKEND_ADSP_MTRACE=y

has to be added to overlay config file to enable this changes.

### CODE REVIEW

**Comment by piotrkmx** [comment]
@nashif Hi Anas, I seems that assigning code owner in CODEOWNERS file is now required for all new files. Please grant proper privileges to my account so I can assign myself.

File: soc/xtensa/intel_adsp/Kconfig
Comment (pjdobrowolski) [inline soc/xtensa/intel_adsp/Kconfig:49]: Previously it was 2 pages of memory, why it is now 4 times more?

Comment (piotrkmx) [reply soc/xtensa/intel_adsp/Kconfig:49]: Hi, there is one page for each slot. According to the documentation, you have to fit 7 slots into this window. There is also one page for descriptors. This gives us 8 pages in total.

File: include/zephyr/drivers/mm/system_mm.h
Comment (jxstelter) [inline include/zephyr/drivers/mm/system_mm.h:286]: The other functions have implementations in mm_drv_intel_adsp_tlb.c and mm_drv_intel_adsp_mtl_tlb.c files. The first one is generic, second specific for MTL TLB case.  Thos one has only MTL TLB specific implementation.

File: include/zephyr/logging/log_backend_adsp_mtrace.h
Comment (jxstelter) [inline include/zephyr/logging/log_backend_adsp_mtrace.h:21]: I am not sure if this header is good place to define PAGE_SIZE. There is another define ADSP_DW_PAGE_SIZE.

Comment (piotrkmx) [reply include/zephyr/logging/log_backend_adsp_mtrace.h:21]: Hi @jxstelter, there was ADSP_DW_PAGE_SIZE in adsp_debug_window.h. I removed that file. But other than that you're right. I followed the example and added page size define to each module by analogy. Not the best solution. I'm thinking about Kconfig option now. Do you have any better suggestions where to put it? I personally haven't found proper existing header file, didn't want to create a separate one just for this define and still want those modules to be as independent from each other as possible to provide some kind of encapsulation.

**Review by softwarecki — CHANGES_REQUESTED** [review]
I noticed an inconsistency in the naming of structures. Some of them use CamelCase when another use a underscore as delimiter. To stay within convention used in the Zephyr project, please use underscore.

File: include/zephyr/logging/log_backend_adsp_mtrace.h
Comment (tmleman) [reply include/zephyr/logging/log_backend_adsp_mtrace.h:21]: What about CONFIG_MM_DRV_PAGE_SIZE?

File: drivers/memory_window/dynamic_slots.h
Comment (kv2019i) [inline drivers/memory_window/dynamic_slots.h:26]: These generic ADSP_ definitions feel a bit out-of-place in header of a "memory_window" driver.

File: drivers/memory_window/dynamic_slots.h
Comment (kv2019i) [inline drivers/memory_window/dynamic_slots.h:28]: In zephyr "__packed" is used to pack structs.

File: drivers/memory_window/dynamic_slots.h
Comment (kv2019i) [inline drivers/memory_window/dynamic_slots.h:29]: I need the Zephyr experts to weigh in on, but I haven't seen much CamelCase in Zephyr codebase.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @piotrkmx , this is quite a big one. I've mostly focused on how this impacts mtrace log. I think the generic approach is good for the debug window management, and replacing the static slot allocation is the right thing to do. I think we'd need some way to support also platforms (like cAVS2.5) where we don't have a TLB that can mirror vmas (i.e. how to keep mtrace working on Tigerlake and other platforms with SOF). It seems this can be added to dynamic_slots init.

There are multiple small style things that will require attention. I wonder if this could be easier to do in smaller chunks, but I'll let other reviewers chime in.

**Comment by marc-hb** [comment]
> Thanks @piotrkmx , this is quite a big one. [...] There are multiple small style things that will require attention. I wonder if this could be easier to do in smaller chunks, but I'll let other reviewers chime in.

It's not just reviewers. The idea of "Continuous Integration" is to validate one change at a time.

Speaking of validation: do submit this Zephyr PR to SOF validation _before_ it's merged in Zephyr. It's as simple as editing sof/west.yml to point at the latest commit here and submit as a draft SOF PR. 



> @nashif Hi Anas, I seems that assigning code owner in CODEOWNERS file is now required for all new files. Please grant proper privileges to my account so I can assign myself.

The current, giant error message is made of many lines of "Unknown owner on line X: make sure @Y exists and has write access to the repository"

This looks unrelated to you or this PR?


From https://docs.zephyrproject.org/latest/project/project_roles.html#maintainers-file

> The MAINTAINERS.yml file shall replace the CODEOWNERS file and will be used for both setting assignees and reviewers.

However the git log shows both still being updated. I don't know.

File: CODEOWNERS
Comment (marc-hb) [inline CODEOWNERS:341]: Listing every single file looks very strange.

File: drivers/memory_window/debug_memory_window.c
Comment (pblaszko) [inline drivers/memory_window/debug_memory_window.c:28]: check error code

File: drivers/memory_window/debug_memory_window.c
Comment (pblaszko) [inline drivers/memory_window/debug_memory_window.c:37]: check error code

File: drivers/memory_window/debug_memory_window.c
Comment (pblaszko) [inline drivers/memory_window/debug_memory_window.c:32]: check error code

File: drivers/memory_window/dynamic_slots.c
Comment (pblaszko) [inline drivers/memory_window/dynamic_slots.c:62]: I think internal infrastructure should use POSIX error codes. Only IPC responses for IPC4 should use positive error codes.
This applies for entire commit.

File: drivers/memory_window/debug_memory_window.c
Comment (pblaszko) [reply drivers/memory_window/debug_memory_window.c:28]: + modify dynamic_slots_init() to return POSIX ec

File: drivers/memory_window/debug_memory_window.c
Comment (pblaszko) [reply drivers/memory_window/debug_memory_window.c:32]: + modify dynamic_slots_map_slot() to return POSIX ec

File: drivers/memory_window/dynamic_slots.h
Comment (piotrkmx) [reply drivers/memory_window/dynamic_slots.h]: No need to. My mistake. Already fixed.

File: soc/xtensa/intel_adsp/Kconfig
Comment (piotrkmx) [reply soc/xtensa/intel_adsp/Kconfig]: I overlooked the possibility to override this values. I moved those changes to soc/xtensa/intel_adsp/ace/Kconfig.defconfig.series. Seems to work fine. Default values can be changed when the time comes.

File: soc/xtensa/intel_adsp/Kconfig
Comment (piotrkmx) [reply soc/xtensa/intel_adsp/Kconfig]: I overlooked the possibility to override this values. I moved those changes to soc/xtensa/intel_adsp/ace/Kconfig.defconfig.series. Seems to work fine. Default values can be changed when the time comes.

File: soc/xtensa/intel_adsp/common/mem_window.c
Comment (piotrkmx) [reply soc/xtensa/intel_adsp/common/mem_window.c]: @kv2019i I'm sorry, but I don't understand. Please elaborate. What do you mean by saying "You later this fix"? Delay? What is indirect about it? Do you just want me to create a separate PR for that?

Comment (kv2019i) [reply soc/xtensa/intel_adsp/common/mem_window.c]: @piotrkmx It's just about where you have this change in the series. You modified this line in two patches in the series (at least the original upload). I was just saying this change should only be in one of the commits.

File: include/zephyr/drivers/mm/system_mm.h
Comment (piotrkmx) [reply include/zephyr/drivers/mm/system_mm.h:286]: I think that this makes this PR even bigger and kind of violates the rule of not committing the changes that are not in use, but I've now added generic support within a separate commit. We can make this a separate PR eventually.

File: drivers/mm/mm_drv_common.c
Comment (softwarecki) [inline drivers/mm/mm_drv_common.c:450]: goto not needed, just return error

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c
Comment (softwarecki) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:691]: why do we need ret2 variable?

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c
Comment (softwarecki) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:729]: Do we need flush cache if any error occurred?

File: include/zephyr/logging/log_backend_adsp_mtrace.h
Comment (piotrkmx) [reply include/zephyr/logging/log_backend_adsp_mtrace.h]: CONFIG_MM_DRV_PAGE_SIZE works fine for most cases, but I believe that log backend mtrace in platforms other than ACE  is supposed to work without mm driver. Anyway I removed PAGE_SIZE from the header file.

File: soc/xtensa/intel_adsp/common/mem_window.c
Comment (piotrkmx) [reply soc/xtensa/intel_adsp/common/mem_window.c]: Oh... now I see. I added a space to it in the last commit instead of adding it in the one with the fix. Most probably because of checkpatch. Fixed.

File: soc/xtensa/intel_adsp/common/include/adsp_debug_window.h
Comment (piotrkmx) [reply soc/xtensa/intel_adsp/common/include/adsp_debug_window.h:1]: To be sincere, I did some research and don't see any other inclusion of adsp_debug_window.h in SOF. The only inclusion was in log backend mtrace in Zephyr and that is now switched to dynamic slots.

File: drivers/memory_window/dynamic_slots.c
Comment (piotrkmx) [reply drivers/memory_window/dynamic_slots.c]: @kv2019i I've added backwards compatibility for legacy platforms using simple observer. Please check if you like the concept.

File: drivers/memory_window/Kconfig
Comment (piotrkmx) [reply drivers/memory_window/Kconfig]: I've added backwards compatibility for legacy platforms using simple observer. Please check if you like the concept.

File: subsys/logging/backends/log_backend_adsp_mtrace.c
Comment (kv2019i) [inline subsys/logging/backends/log_backend_adsp_mtrace.c:143]: Hmm, this would create a 4K memcpy for each log addition. This seems a bit too much overhead. Couldn't we have an option to get the buffer address from dynamic_slots.h, allowing to directly modify the slot memory. I get the benefit when mapping data that is anyways stored elsewhere and the mm remapping allows to avoid to same memory (and writes), but for this log data, there's no savings here and direct mode should be doable as well.

File: drivers/telemetry/telemetry_intel.h
Comment (kv2019i) [inline drivers/telemetry/telemetry_intel.h:8]: @nashif @MaureenHelm @teburd This general concept of adding hw/vendor specific telemetry interface requires feedback. I don't see any existing telemetry infra in Zephyr now, so starting with a vendor specific interface seems pragmatic start, but requires attention anyways. FYI @mwasko  @marcinszkudlinski

File: subsys/logging/backends/log_backend_adsp_mtrace.c
Comment (kv2019i) [reply subsys/logging/backends/log_backend_adsp_mtrace.c:143]: And another issue is that the host will also modify the first bits of this slot and this observer construct will basicly not reflect the data host wrote (and the memcpy potentially overwrites anything the host writes).

**Review by kv2019i — CHANGES_REQUESTED** [review]
I did a test run of the observer-functionality, and I'm afraid it didn't quite work (on a TGL machine). I added some comments inline. I wonder if we could have a "direct modify" mode to dynamic_slots.h which could be used on legacy platforms. Pinged a few more folks to get review attention to the telemetry and memory_window interfaces.

File: subsys/logging/backends/log_backend_adsp_mtrace.c
Comment (piotrkmx) [reply subsys/logging/backends/log_backend_adsp_mtrace.c]: This part of code has been removed.

File: subsys/logging/backends/log_backend_adsp_mtrace.c
Comment (piotrkmx) [reply subsys/logging/backends/log_backend_adsp_mtrace.c]: @kv2019i Hi, please check the new version. Dynamic slots allocation is still available for both platforms, but now you can  choose if you want to use TLB with mirroring or write directly to the memory as you were before for each slot separately. This should not affect legacy platform like TGL.

File: drivers/telemetry/telemetry_intel.h
Comment (piotrkmx) [reply drivers/telemetry/telemetry_intel.h:8]: Hi, what I've put here is only a telemetry base structure, to show the purpose of this PR and that it works. This structure cannot be modified anyway, as it is supposed to be seen by host like it is. I've tried to make the last two commits a separate PR, so we can discuss it there, but there is one dependency that cannot be overcome until this PR is merged and it has to go together. So my suggestion is to postpone this this discussion till next PR's with proper API's will be ready - the ones that actually fill the data.

Comment (mwasko) [reply drivers/telemetry/telemetry_intel.h:8]: @kv2019i as you metioned, the telemetry concept is unique for Intel configuration for now, so I recommend to continue with integration of current solution as it is in the first step. There is no telemetry interface that we could align with today.

@nashif, @MaureenHelm, @teburd it would be probably ask to you to prepare generic Telemetry interface proposition using Intel implementation as an input. Once the generic interface definition will be ready we can apply it to Intel telemetry driver.

File: subsys/logging/backends/log_backend_adsp_mtrace.c
Comment (kv2019i) [reply subsys/logging/backends/log_backend_adsp_mtrace.c]: Thanks @piotrkmx! Will check this and comment back here.

**Comment by kv2019i** [comment]
@piotrkmx  I tested the latest revision today. Code seems correct now and I was assuming it would work, but mtrace is still not producing any logs on cAVS2.5. I can see the DSP write pointer is not getting updated at all:

[3897079.472493] sof-audio-pci-intel-tgl 0000:00:1f.3: core0, host read: 0x0, dsp write: 0x9010e0c
[3897079.472499] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx done : 0x1b060000|0x0: GLB_NOTIFICATION|LOG_BUFFER_STATUS
[3897080.496472] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx      : 0x1b060000|0x0: GLB_NOTIFICATION|LOG_BUFFER_STATUS
[3897080.496493] sof-audio-pci-intel-tgl 0000:00:1f.3: core0, host read: 0x0, dsp write: 0x9010e0c
[3897080.496499] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx done : 0x1b060000|0x0: GLB_NOTIFICATION|LOG_BUFFER_STATUS
[3897081.520485] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc rx      : 0x1b060000|0x0: GLB_NOTIFICATION|LOG_BUFFER_STATUS
[3897081.520508] sof-audio-pci-intel-tgl 0000:00:1f.3: core0, host read: 0x0, dsp write: 0x9010e0c

The "dsp write" value seems bogus, it should reflect the position within the 4KiB page where DSP has written to. 

Potential issue -- compiler optimizations (I see at least some volatile attributes dropped in the patches), or the initialization is not called. Debugging these is a bit slow as without mtrace, well, you don't have any logs what the DSP is doing.

**Comment by piotrkmx** [comment]
> @piotrkmx I tested the latest revision today. Code seems correct now and I was assuming it would work, but mtrace is still not producing any logs on cAVS2.5.

I was not able to test it on on cavs platform yet (only ace), I'll try to do this as soon as I manage to fix warnings from SOF CI. But just to be sure, did you remember to enable it in SOF board config file?

CONFIG_DYNAMIC_SLOTS=y
CONFIG_DEBUG_MEMORY_WINDOW=y
CONFIG_LOG_BACKEND_ADSP_MTRACE=y

File: drivers/memory_window/dynamic_slots.c
Comment (lyakh) [inline drivers/memory_window/dynamic_slots.c:24]: `arch_xtensa_cached_ptr()` returns a pointer from `__sparse_cache` address space, so here you want to remove it instead of enforcing it again:
```
uint32_t *virt_buf = (__sparse_force uint32_t *)
		arch_xtensa_cached_ptr((void *)(config->mem_base));
```

File: drivers/memory_window/dynamic_slots.c
Comment (lyakh) [inline drivers/memory_window/dynamic_slots.c:38]: same here

File: drivers/memory_window/dynamic_slots.c
Comment (lyakh) [inline drivers/memory_window/dynamic_slots.c:45]: here `buf` must be `__sparse_cache` or must be type-cast to it. So, maybe in this patch it would be good to declare `uint32_t __sparse_cache *buf`. But check where else it is used, maybe you'll end up casting it back to the default address space too often

File: drivers/memory_window/dynamic_slots.c
Comment (lyakh) [inline drivers/memory_window/dynamic_slots.c:109]: same

File: drivers/memory_window/dynamic_slots.c
Comment (lyakh) [inline drivers/memory_window/dynamic_slots.c:112]: just like above

File: drivers/memory_window/dynamic_slots.c
Comment (lyakh) [inline drivers/memory_window/dynamic_slots.c:155]: here too

File: drivers/memory_window/dynamic_slots.c
Comment (lyakh) [inline drivers/memory_window/dynamic_slots.c:158]: and here

File: drivers/telemetry/telemetry_intel.c
Comment (lyakh) [inline drivers/telemetry/telemetry_intel.c:23]: `z_soc_cached_ptr()` returns `__sparse_cache` so you have to match them

**Comment by kv2019i** [comment]
@piotrkmx wrote:
> fix warnings from SOF CI. But just to be sure, did you remember to enable it in SOF board config file?
> 
> CONFIG_DYNAMIC_SLOTS=y CONFIG_DEBUG_MEMORY_WINDOW=y CONFIG_LOG_BACKEND_ADSP_MTRACE=y

Good point. I had not modified the board file, but looking at the build .config, all the above are defined, so it seems the dependencies bring them in (or are enabled by default).

**Pull request closed without merging**
